builder/dockerfile/internals.go

```diff
@@ -298,16 +297,30 @@ func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
 	return &builder.HashedFileInfo{FileInfo: builder.PathFileInfo{FileInfo: tmpFileSt, FilePath: tmpFileName}, FileHash: hash}, nil
 }
 
+var windowsBlacklist = append([]string{}, "c:\\", "c:\\windows")
+
 func (b *Builder) calcCopyInfo(cmdName, origPath string, allowLocalDecompression, allowWildcards bool, imageSource *imageMount) ([]copyInfo, error) {
 
 	// Work in daemon-specific OS filepath semantics
 	origPath = filepath.FromSlash(origPath)
-
 	// validate windows paths from other images
 	if imageSource != nil && runtime.GOOS == "windows" {
-		forbid := regexp.MustCompile("(?i)^" + string(os.PathSeparator) + "?(windows(" + string(os.PathSeparator) + ".+)?)?$")
-		if p := filepath.Clean(origPath); p == "." || forbid.MatchString(p) {
-			return nil, errors.Errorf("copy from %s is not allowed on windows", origPath)
+		p := strings.ToLower(filepath.Clean(origPath))
+		if !filepath.IsAbs(p) {
```

Comment:
Not on a Windows machine, but instead of manually making this path absolute, would this work?

```go
p := strings.ToLower(filepath.Abs(origPath))
for _, forbidden := range windowsBlacklist {
if p == forbidden {
	return nil, errors.New("copy from c:\\ or c:\\windows is not allowed on windows")
}
```

To prevent the loop, perhaps;

```go
windowsBlacklist := map[string]interface{}{
  "c:\\" : true,
  "c:\\windows" : true,
}

if _, ok := windowsBlacklist[p]; ok {
	return nil, errors.New("copy from "+p+" is not allowed on windows")
}
```